from PyQt6.QtWidgets import QMessageBox, QTableWidgetItem, QHeaderView, QDialog, QVBoxLayout, QLabel
from PyQt6.QtCore import Qt
from Event_Check_in_Management.ui.MainWindow import Ui_MainWindow
from Event_Check_in_Management.ui.EventDialogEx import EventDialogEx
from Event_Check_in_Management.ui.AttendeeDialogEx import AttendeeDialogEx
from Event_Check_in_Management.ui.RegistrationDialogEx import RegistrationDialogEx
from Event_Check_in_Management.models.events import Events
from Event_Check_in_Management.models.event import Event
from Event_Check_in_Management.models.attendees import Attendees
from Event_Check_in_Management.models.attendee import Attendee
from Event_Check_in_Management.models.registrations import Registrations
from Event_Check_in_Management.models.registration import Registration
from datetime import datetime
import uuid

try:
    import qrcode
    from PyQt6.QtGui import QPixmap, QImage
    from io import BytesIO
    QR_AVAILABLE = True
except ImportError:
    QR_AVAILABLE = False

class MainWindowEx(Ui_MainWindow):
    def setupUi(self, MainWindow):
        super().setupUi(MainWindow)
        self.MainWindow = MainWindow
        self.setupSignalAndSlot()
        self.load_initial_data()
        self.apply_stylesheet()
    
    def showWindow(self):
        self.MainWindow.show()
    
    def setupSignalAndSlot(self):
        """K·∫øt n·ªëi c√°c signals v·ªõi slots"""
        # Tab S·ª± Ki·ªán
        self.btnAddEvent.clicked.connect(self.add_event)
        self.btnViewEvent.clicked.connect(self.view_event_details)
        self.btnEditEvent.clicked.connect(self.edit_event)
        self.btnDeleteEvent.clicked.connect(self.delete_event)
        self.btnRefreshEvent.clicked.connect(self.load_events)
        
        # Tab Ng∆∞·ªùi Tham D·ª±
        self.attendeeSearch.textChanged.connect(self.search_attendees)
        self.btnAddAttendee.clicked.connect(self.add_attendee)
        self.btnEditAttendee.clicked.connect(self.edit_attendee)
        self.btnDeleteAttendee.clicked.connect(self.delete_attendee)
        self.btnRefreshAttendee.clicked.connect(self.load_attendees)
        
        # Tab ƒêƒÉng K√Ω
        self.eventCombo.currentIndexChanged.connect(self.load_registrations)
        self.btnRegisterAttendee.clicked.connect(self.register_attendee)
        self.btnGenerateQR.clicked.connect(self.generate_qr_code)
        self.btnCancelRegistration.clicked.connect(self.cancel_registration)
        self.btnRefreshRegistration.clicked.connect(self.load_registrations)
        
        # Tab Check-in
        self.checkinEventCombo.currentIndexChanged.connect(self.load_checkin_stats)
        self.btnCheckin.clicked.connect(self.perform_checkin)
        self.btnRefreshCheckin.clicked.connect(self.load_checkin_stats)
        
        # Thi·∫øt l·∫≠p header cho c√°c table
        self.eventTable.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
        self.attendeeTable.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
        self.registrationTable.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
        self.checkinTable.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
    
    def load_initial_data(self):
        """Load d·ªØ li·ªáu ban ƒë·∫ßu"""
        self.load_events()
        self.load_attendees()
        self.load_event_combo()
        self.load_checkin_event_combo()
    
    # ===== QU·∫¢N L√ù S·ª∞ KI·ªÜN =====
    def load_events(self):
        """Load danh s√°ch s·ª± ki·ªán"""
        events = Events()
        events.import_json("datasets/events.json")
        
        self.eventTable.setRowCount(len(events.list))
        
        for row, event in enumerate(events.list):
            self.eventTable.setItem(row, 0, QTableWidgetItem(event.EventId))
            self.eventTable.setItem(row, 1, QTableWidgetItem(event.EventName))
            self.eventTable.setItem(row, 2, QTableWidgetItem(event.EventDate))
            self.eventTable.setItem(row, 3, QTableWidgetItem(event.EventTime))
            self.eventTable.setItem(row, 4, QTableWidgetItem(event.Location))
            self.eventTable.setItem(row, 5, QTableWidgetItem(event.Description if event.Description else ""))
    
    def add_event(self):
        """Th√™m s·ª± ki·ªán m·ªõi"""
        dialog = EventDialogEx(self.MainWindow)
        if dialog.exec() == QDialog.DialogCode.Accepted:
            data = dialog.get_data()
            
            # T·∫°o event m·ªõi
            event = Event()
            event.EventId = "evt_" + str(uuid.uuid4())[:8]
            event.EventName = data['name']
            event.EventDate = data['date']
            event.EventTime = data['time']
            event.Location = data['location']
            event.Description = data['description']
            
            # L∆∞u v√†o file
            events = Events()
            events.import_json("datasets/events.json")
            events.add_item(event)
            events.export_json("datasets/events.json")
            
            QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", "ƒê√£ th√™m s·ª± ki·ªán m·ªõi!")
            self.load_events()
            self.load_event_combo()
            self.load_checkin_event_combo()
    
    def edit_event(self):
        """S·ª≠a s·ª± ki·ªán"""
        current_row = self.eventTable.currentRow()
        if current_row < 0:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng ch·ªçn s·ª± ki·ªán c·∫ßn s·ª≠a!")
            return
        
        event_id = self.eventTable.item(current_row, 0).text()
        
        # T√¨m event
        events = Events()
        events.import_json("datasets/events.json")
        event = events.find_event(event_id)
        
        if event:
            dialog = EventDialogEx(self.MainWindow, event)
            if dialog.exec() == QDialog.DialogCode.Accepted:
                data = dialog.get_data()
                
                event.EventName = data['name']
                event.EventDate = data['date']
                event.EventTime = data['time']
                event.Location = data['location']
                event.Description = data['description']
                
                events.export_json("datasets/events.json")
                
                QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", "ƒê√£ c·∫≠p nh·∫≠t s·ª± ki·ªán!")
                self.load_events()
                self.load_event_combo()
                self.load_checkin_event_combo()
    
    def delete_event(self):
        """X√≥a s·ª± ki·ªán"""
        current_row = self.eventTable.currentRow()
        if current_row < 0:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng ch·ªçn s·ª± ki·ªán c·∫ßn x√≥a!")
            return
        
        reply = QMessageBox.question(
            self.MainWindow, 
            "X√°c Nh·∫≠n",
            "B·∫°n c√≥ ch·∫Øc mu·ªën x√≥a s·ª± ki·ªán n√†y?\nT·∫•t c·∫£ ƒëƒÉng k√Ω li√™n quan s·∫Ω b·ªã x√≥a!",
            QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No
        )
        
        if reply == QMessageBox.StandardButton.Yes:
            event_id = self.eventTable.item(current_row, 0).text()
            
            # X√≥a event
            events = Events()
            events.import_json("datasets/events.json")
            events.delete_event(event_id)
            events.export_json("datasets/events.json")
            
            # X√≥a c√°c registrations li√™n quan
            registrations = Registrations()
            registrations.import_json("datasets/registrations.json")
            registrations.list = [r for r in registrations.list if r.EventId != event_id]
            registrations.export_json("datasets/registrations.json")
            
            QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", "ƒê√£ x√≥a s·ª± ki·ªán!")
            self.load_events()
            self.load_event_combo()
            self.load_checkin_event_combo()
    
    def view_event_details(self):
        """Xem chi ti·∫øt s·ª± ki·ªán"""
        current_row = self.eventTable.currentRow()
        if current_row < 0:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng ch·ªçn s·ª± ki·ªán!")
            return
        
        event_id = self.eventTable.item(current_row, 0).text()
        
        events = Events()
        events.import_json("datasets/events.json")
        event = events.find_event(event_id)
        
        if event:
            registrations = Registrations()
            registrations.import_json("datasets/registrations.json")
            
            total_reg = registrations.count_registered_by_event(event_id)
            total_checkin = registrations.count_checkedin_by_event(event_id)
            
            details = f"""
            <h2>{event.EventName}</h2>
            <p><b>üìÖ Ng√†y:</b> {event.EventDate}</p>
            <p><b>üïê Gi·ªù:</b> {event.EventTime}</p>
            <p><b>üìç ƒê·ªãa ƒëi·ªÉm:</b> {event.Location}</p>
            <p><b>üìù M√¥ t·∫£:</b> {event.Description if event.Description else 'Kh√¥ng c√≥'}</p>
            <hr>
            <p><b>üë• T·ªïng ƒëƒÉng k√Ω:</b> {total_reg}</p>
            <p><b>‚úÖ ƒê√£ check-in:</b> {total_checkin}</p>
            """
            
            QMessageBox.information(self.MainWindow, "Chi Ti·∫øt S·ª± Ki·ªán", details)
    
    # ===== QU·∫¢N L√ù NG∆Ø·ªúI THAM D·ª∞ =====
    def load_attendees(self):
        """Load danh s√°ch ng∆∞·ªùi tham d·ª±"""
        attendees = Attendees()
        attendees.import_json("datasets/attendees.json")
        
        self.attendeeTable.setRowCount(len(attendees.list))
        
        for row, attendee in enumerate(attendees.list):
            self.attendeeTable.setItem(row, 0, QTableWidgetItem(attendee.AttendeeId))
            self.attendeeTable.setItem(row, 1, QTableWidgetItem(attendee.Name))
            self.attendeeTable.setItem(row, 2, QTableWidgetItem(attendee.Email))
            self.attendeeTable.setItem(row, 3, QTableWidgetItem(attendee.Phone if attendee.Phone else ""))
            self.attendeeTable.setItem(row, 4, QTableWidgetItem(attendee.Organization if attendee.Organization else ""))
            self.attendeeTable.setItem(row, 5, QTableWidgetItem(attendee.Position if attendee.Position else ""))
    
    def search_attendees(self):
        """T√¨m ki·∫øm ng∆∞·ªùi tham d·ª±"""
        keyword = self.attendeeSearch.text().strip()
        
        if not keyword:
            self.load_attendees()
            return
        
        attendees = Attendees()
        attendees.import_json("datasets/attendees.json")
        results = attendees.search_attendees(keyword)
        
        self.attendeeTable.setRowCount(len(results))
        
        for row, attendee in enumerate(results):
            self.attendeeTable.setItem(row, 0, QTableWidgetItem(attendee.AttendeeId))
            self.attendeeTable.setItem(row, 1, QTableWidgetItem(attendee.Name))
            self.attendeeTable.setItem(row, 2, QTableWidgetItem(attendee.Email))
            self.attendeeTable.setItem(row, 3, QTableWidgetItem(attendee.Phone if attendee.Phone else ""))
            self.attendeeTable.setItem(row, 4, QTableWidgetItem(attendee.Organization if attendee.Organization else ""))
            self.attendeeTable.setItem(row, 5, QTableWidgetItem(attendee.Position if attendee.Position else ""))
    
    def add_attendee(self):
        """Th√™m ng∆∞·ªùi tham d·ª± m·ªõi"""
        dialog = AttendeeDialogEx(self.MainWindow)
        if dialog.exec() == QDialog.DialogCode.Accepted:
            data = dialog.get_data()
            
            # Ki·ªÉm tra email tr√πng
            attendees = Attendees()
            attendees.import_json("datasets/attendees.json")
            
            for att in attendees.list:
                if att.Email.lower() == data['email'].lower():
                    QMessageBox.warning(self.MainWindow, "L·ªói", "Email ƒë√£ t·ªìn t·∫°i!")
                    return
            
            # T·∫°o attendee m·ªõi
            attendee = Attendee()
            attendee.AttendeeId = "att_" + str(uuid.uuid4())[:8]
            attendee.Name = data['name']
            attendee.Email = data['email']
            attendee.Phone = data['phone']
            attendee.Organization = data['organization']
            attendee.Position = data['position']
            
            attendees.add_item(attendee)
            attendees.export_json("datasets/attendees.json")
            
            QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", "ƒê√£ th√™m ng∆∞·ªùi tham d·ª±!")
            self.load_attendees()
    
    def edit_attendee(self):
        """S·ª≠a ng∆∞·ªùi tham d·ª±"""
        current_row = self.attendeeTable.currentRow()
        if current_row < 0:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng ch·ªçn ng∆∞·ªùi tham d·ª± c·∫ßn s·ª≠a!")
            return
        
        attendee_id = self.attendeeTable.item(current_row, 0).text()
        
        attendees = Attendees()
        attendees.import_json("datasets/attendees.json")
        attendee = attendees.find_attendee(attendee_id)
        
        if attendee:
            dialog = AttendeeDialogEx(self.MainWindow, attendee)
            if dialog.exec() == QDialog.DialogCode.Accepted:
                data = dialog.get_data()
                
                # Ki·ªÉm tra email tr√πng (tr·ª´ ch√≠nh n√≥)
                for att in attendees.list:
                    if att.Email.lower() == data['email'].lower() and att.AttendeeId != attendee_id:
                        QMessageBox.warning(self.MainWindow, "L·ªói", "Email ƒë√£ t·ªìn t·∫°i!")
                        return
                
                attendee.Name = data['name']
                attendee.Email = data['email']
                attendee.Phone = data['phone']
                attendee.Organization = data['organization']
                attendee.Position = data['position']
                
                attendees.export_json("datasets/attendees.json")
                
                QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", "ƒê√£ c·∫≠p nh·∫≠t ng∆∞·ªùi tham d·ª±!")
                self.load_attendees()
    
    def delete_attendee(self):
        """X√≥a ng∆∞·ªùi tham d·ª±"""
        current_row = self.attendeeTable.currentRow()
        if current_row < 0:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng ch·ªçn ng∆∞·ªùi tham d·ª± c·∫ßn x√≥a!")
            return
        
        reply = QMessageBox.question(
            self.MainWindow,
            "X√°c Nh·∫≠n",
            "B·∫°n c√≥ ch·∫Øc mu·ªën x√≥a ng∆∞·ªùi tham d·ª± n√†y?\nT·∫•t c·∫£ ƒëƒÉng k√Ω li√™n quan s·∫Ω b·ªã x√≥a!",
            QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No
        )
        
        if reply == QMessageBox.StandardButton.Yes:
            attendee_id = self.attendeeTable.item(current_row, 0).text()
            
            attendees = Attendees()
            attendees.import_json("datasets/attendees.json")
            attendees.delete_attendee(attendee_id)
            attendees.export_json("datasets/attendees.json")
            
            # X√≥a c√°c registrations li√™n quan
            registrations = Registrations()
            registrations.import_json("datasets/registrations.json")
            registrations.list = [r for r in registrations.list if r.AttendeeId != attendee_id]
            registrations.export_json("datasets/registrations.json")
            
            QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", "ƒê√£ x√≥a ng∆∞·ªùi tham d·ª±!")
            self.load_attendees()
    
    # ===== QU·∫¢N L√ù ƒêƒÇNG K√ù =====
    def load_event_combo(self):
        """Load combo box s·ª± ki·ªán"""
        self.eventCombo.clear()
        events = Events()
        events.import_json("datasets/events.json")
        
        for event in events.list:
            self.eventCombo.addItem(f"{event.EventName} - {event.EventDate}", event.EventId)
    
    def load_registrations(self):
        """Load danh s√°ch ƒëƒÉng k√Ω"""
        if self.eventCombo.count() == 0:
            self.registrationTable.setRowCount(0)
            return
        
        event_id = self.eventCombo.currentData()
        if not event_id:
            return
        
        registrations = Registrations()
        registrations.import_json("datasets/registrations.json")
        regs = registrations.get_registrations_by_event(event_id)
        
        attendees = Attendees()
        attendees.import_json("datasets/attendees.json")
        
        self.registrationTable.setRowCount(len(regs))
        
        for row, reg in enumerate(regs):
            attendee = attendees.find_attendee(reg.AttendeeId)
            if attendee:
                self.registrationTable.setItem(row, 0, QTableWidgetItem(reg.RegistrationId))
                self.registrationTable.setItem(row, 1, QTableWidgetItem(attendee.Name))
                self.registrationTable.setItem(row, 2, QTableWidgetItem(attendee.Email))
                self.registrationTable.setItem(row, 3, QTableWidgetItem(attendee.Organization if attendee.Organization else ""))
                self.registrationTable.setItem(row, 4, QTableWidgetItem(reg.RegistrationDate))
                self.registrationTable.setItem(row, 5, QTableWidgetItem(reg.Status))
    
    def register_attendee(self):
        """ƒêƒÉng k√Ω ng∆∞·ªùi tham d·ª± v√†o s·ª± ki·ªán"""
        if self.eventCombo.count() == 0:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Kh√¥ng c√≥ s·ª± ki·ªán n√†o!")
            return
        
        event_id = self.eventCombo.currentData()
        if not event_id:
            return
        
        dialog = RegistrationDialogEx(self.MainWindow)
        if dialog.exec() == QDialog.DialogCode.Accepted:
            attendee_id = dialog.get_selected_attendee_id()
            
            if not attendee_id:
                QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng ch·ªçn ng∆∞·ªùi tham d·ª±!")
                return
            
            # Ki·ªÉm tra ƒë√£ ƒëƒÉng k√Ω ch∆∞a
            registrations = Registrations()
            registrations.import_json("datasets/registrations.json")
            
            existing = registrations.find_registration_by_event_attendee(event_id, attendee_id)
            if existing:
                QMessageBox.warning(self.MainWindow, "L·ªói", "Ng∆∞·ªùi n√†y ƒë√£ ƒëƒÉng k√Ω s·ª± ki·ªán!")
                return
            
            # T·∫°o registration m·ªõi
            reg = Registration()
            reg.RegistrationId = str(uuid.uuid4())[:8].upper()
            reg.EventId = event_id
            reg.AttendeeId = attendee_id
            reg.RegistrationDate = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            reg.Status = "ƒê√£ ƒëƒÉng k√Ω"
            
            registrations.add_item(reg)
            registrations.export_json("datasets/registrations.json")
            
            QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", f"ƒê√£ ƒëƒÉng k√Ω!\nM√£ ƒëƒÉng k√Ω: {reg.RegistrationId}")
            self.load_registrations()
    
    def cancel_registration(self):
        """H·ªßy ƒëƒÉng k√Ω"""
        current_row = self.registrationTable.currentRow()
        if current_row < 0:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng ch·ªçn ƒëƒÉng k√Ω c·∫ßn h·ªßy!")
            return
        
        reply = QMessageBox.question(
            self.MainWindow,
            "X√°c Nh·∫≠n",
            "B·∫°n c√≥ ch·∫Øc mu·ªën h·ªßy ƒëƒÉng k√Ω n√†y?",
            QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No
        )
        
        if reply == QMessageBox.StandardButton.Yes:
            reg_id = self.registrationTable.item(current_row, 0).text()
            
            registrations = Registrations()
            registrations.import_json("datasets/registrations.json")
            registrations.delete_registration(reg_id)
            registrations.export_json("datasets/registrations.json")
            
            QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", "ƒê√£ h·ªßy ƒëƒÉng k√Ω!")
            self.load_registrations()
    
    def generate_qr_code(self):
        """T·∫°o m√£ QR cho ƒëƒÉng k√Ω"""
        if not QR_AVAILABLE:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Ch∆∞a c√†i ƒë·∫∑t th∆∞ vi·ªán qrcode!\nC√†i ƒë·∫∑t: pip install qrcode[pil]")
            return
        
        current_row = self.registrationTable.currentRow()
        if current_row < 0:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng ch·ªçn ƒëƒÉng k√Ω!")
            return
        
        reg_id = self.registrationTable.item(current_row, 0).text()
        attendee_name = self.registrationTable.item(current_row, 1).text()
        
        # T·∫°o QR code
        qr = qrcode.QRCode(version=1, box_size=10, border=5)
        qr.add_data(reg_id)
        qr.make(fit=True)
        
        img = qr.make_image(fill_color="black", back_color="white")
        
        # Chuy·ªÉn ƒë·ªïi sang QPixmap
        buffer = BytesIO()
        img.save(buffer, format='PNG')
        buffer.seek(0)
        
        qimage = QImage()
        qimage.loadFromData(buffer.read())
        pixmap = QPixmap.fromImage(qimage)
        
        # Hi·ªÉn th·ªã dialog v·ªõi QR code
        qr_dialog = QDialog(self.MainWindow)
        qr_dialog.setWindowTitle("M√£ QR Check-in")
        qr_dialog.resize(400, 450)
        
        qr_layout = QVBoxLayout()
        
        qr_label = QLabel()
        qr_label.setPixmap(pixmap)
        qr_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        
        code_label = QLabel(f"<h3>{attendee_name}</h3><h2>M√£: {reg_id}</h2>")
        code_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        
        qr_layout.addWidget(qr_label)
        qr_layout.addWidget(code_label)
        
        qr_dialog.setLayout(qr_layout)
        qr_dialog.exec()
    
    # ===== CHECK-IN =====
    def load_checkin_event_combo(self):
        """Load combo box s·ª± ki·ªán cho check-in"""
        self.checkinEventCombo.clear()
        events = Events()
        events.import_json("datasets/events.json")
        
        for event in events.list:
            self.checkinEventCombo.addItem(f"{event.EventName} - {event.EventDate}", event.EventId)
    
    def load_checkin_stats(self):
        """Load th·ªëng k√™ check-in"""
        if self.checkinEventCombo.count() == 0:
            self.totalRegisteredLabel.setText("0")
            self.totalCheckedinLabel.setText("0")
            self.checkinTable.setRowCount(0)
            return
        
        event_id = self.checkinEventCombo.currentData()
        if not event_id:
            return
        
        registrations = Registrations()
        registrations.import_json("datasets/registrations.json")
        
        total_reg = registrations.count_registered_by_event(event_id)
        total_checkin = registrations.count_checkedin_by_event(event_id)
        
        self.totalRegisteredLabel.setText(str(total_reg))
        self.totalCheckedinLabel.setText(str(total_checkin))
        
        # Load danh s√°ch ƒë√£ check-in
        regs = registrations.get_registrations_by_event(event_id)
        checkedin = [r for r in regs if r.Status == "ƒê√£ check-in"]
        
        attendees = Attendees()
        attendees.import_json("datasets/attendees.json")
        
        self.checkinTable.setRowCount(len(checkedin))
        
        for row, reg in enumerate(checkedin):
            attendee = attendees.find_attendee(reg.AttendeeId)
            if attendee:
                self.checkinTable.setItem(row, 0, QTableWidgetItem(attendee.Name))
                self.checkinTable.setItem(row, 1, QTableWidgetItem(attendee.Email))
                self.checkinTable.setItem(row, 2, QTableWidgetItem(attendee.Organization if attendee.Organization else ""))
                self.checkinTable.setItem(row, 3, QTableWidgetItem(reg.CheckinTime if reg.CheckinTime else ""))
                self.checkinTable.setItem(row, 4, QTableWidgetItem(reg.RegistrationId))
    
    def perform_checkin(self):
        """Th·ª±c hi·ªán check-in"""
        code = self.checkinCode.text().strip().upper()
        
        if not code:
            QMessageBox.warning(self.MainWindow, "L·ªói", "Vui l√≤ng nh·∫≠p m√£ ƒëƒÉng k√Ω!")
            return
        
        registrations = Registrations()
        registrations.import_json("datasets/registrations.json")
        
        success, message = registrations.checkin(code)
        
        if success:
            registrations.export_json("datasets/registrations.json")
            QMessageBox.information(self.MainWindow, "Th√†nh C√¥ng", message)
            self.checkinCode.clear()
            self.load_checkin_stats()
        else:
            QMessageBox.warning(self.MainWindow, "L·ªói", message)
    
    def apply_stylesheet(self):
        """√Åp d·ª•ng stylesheet cho ·ª©ng d·ª•ng"""
        self.MainWindow.setStyleSheet("""
            QMainWindow {
                background-color: #ecf0f1;
            }
            QPushButton {
                background-color: #3498db;
                color: white;
                border: none;
                padding: 8px 15px;
                border-radius: 4px;
                font-size: 12px;
            }
            QPushButton:hover {
                background-color: #2980b9;
            }
            QTableWidget {
                background-color: white;
                border: 1px solid #bdc3c7;
                border-radius: 4px;
            }
            QHeaderView::section {
                background-color: #34495e;
                color: white;
                padding: 8px;
                border: none;
                font-weight: bold;
            }
            QLineEdit, QTextEdit, QComboBox, QDateEdit, QTimeEdit {
                padding: 6px;
                border: 1px solid #bdc3c7;
                border-radius: 4px;
                background-color: white;
            }
            QGroupBox {
                border: 2px solid #bdc3c7;
                border-radius: 5px;
                margin-top: 10px;
                font-weight: bold;
                padding: 15px;
            }
            QGroupBox::title {
                subcontrol-origin: margin;
                subcontrol-position: top left;
                padding: 5px 10px;
                background-color: white;
            }
        """)
